languages/rust/src/polar.rs

```diff
@@ -0,0 +1,127 @@
+        }
+
+        Self {
+            host: Arc::new(Mutex::new(host)),
+            inner,
+        }
+    }
+
+    pub fn clear(&mut self) {
+        *self = Self::new();
+    }
+
+    fn check_messages(&mut self) {
+        while let Some(msg) = self.inner.next_message() {
+            match msg.kind {
+                polar_core::types::MessageKind::Print => tracing::trace!("{}", msg.msg),
+                polar_core::types::MessageKind::Warning => tracing::warn!("{}", msg.msg),
+            }
+        }
+    }
+
+    fn check_inline_queries(&mut self) -> anyhow::Result<()> {
+        while let Some(q) = self.inner.next_inline_query(false) {
+            let query = crate::query::Query::new(q, self.host.clone());
+            match query.collect::<anyhow::Result<Vec<_>>>() {
+                Ok(v) if !v.is_empty() => continue,
+                Ok(_) => anyhow::bail!("inline query result was false"),
+                Err(e) => {
+                    anyhow::bail!("error in inline query: {}", e);
+                }
+            }
+        }
+        self.check_messages();
+        Ok(())
+    }
+
+    pub fn load_file(&mut self, file: &str) -> anyhow::Result<()> {
+        let mut f = File::open(&file)?;
+        let mut policy = String::new();
+        f.read_to_string(&mut policy)?;
+        self.inner.load(&policy, Some(file.to_string()))?;
+        self.check_inline_queries()
+    }
+
+    pub fn load_str(&mut self, s: &str) -> anyhow::Result<()> {
+        self.inner.load(s, None)?;
+        self.check_inline_queries()
+    }
+
+    pub fn query(&mut self, s: &str) -> anyhow::Result<crate::query::Query> {
+        let query = self.inner.new_query(s, false)?;
+        self.check_messages();
+        let query = crate::query::Query::new(query, self.host.clone());
+        Ok(query)
+    }
+
+    pub fn query_rule<'a>(
+        &mut self,
+        name: &str,
+        args: impl IntoIterator<Item = &'a dyn crate::host::ToPolar>,
+    ) -> anyhow::Result<crate::query::Query> {
+        let args = args
+            .into_iter()
+            .map(|arg| arg.to_polar(&mut self.host.lock().unwrap()))
+            .collect();
+        let query_value = polar_core::types::Value::Call(polar_core::types::Predicate {
+            name: polar_core::types::Symbol(name.to_string()),
+            args,
+        });
+        let query_term = polar_core::types::Term::new_from_ffi(query_value);
+        let query = self.inner.new_query_from_term(query_term, false);
+        self.check_messages();
+        let query = crate::query::Query::new(query, self.host.clone());
+        Ok(query)
+    }
+
+    pub fn register_class(&mut self, class: crate::host::Class) -> anyhow::Result<()> {
+        let mut host = self.host.lock().unwrap();
+        let name = class.name.clone();
+        let name = polar_core::types::Symbol(name);
+        let class_term = host.cache_class(class, name.clone());
+        self.inner.register_constant(name, class_term);
```

Comment:
Sounds worth a try. I think this works if we do your other suggestion of implementing `ToPolar` for `Class`. The difference with other languages though is that we cache a `Class` but register an `Instance` of a `Class<Type>`. I think it will work as you describe though

Reply:
This worked nicely I think